crud_for_code_review.py

```diff
@@ -0,0 +1,546 @@
+    #total number of genres
+    genre_count = len(genres)
+    print(genre_count)
+
+    return (max_genre, max_genre_artists, genre_count)
+
+
+def get_num_artists(user_id):
+    """Gets count of user artists (50 are not always returned)"""
+    user = User.query.get(user_id)
+
+    num_artists = len(user.artists)
+
+    return num_artists
+
+
+def get_user_artists(user_id):
+    """Returns Artist objects for user"""
+
+    user = User.query.get(user_id)
+
+    return (user.artists)
+
+
+def get_user_artist_ids(user_id):
+    """Returns list of artist ids by user artists"""
+
+    user = User.query.get(user_id)
+
+    artist_ids = []
+    for artist in user.artists:
+        artist_ids.append(artist.artist_id)
+
+    return artist_ids
+
+
+def check_user_api_type(user_id, api_type):
+    """Check if a time range or playlist id is in a user's artists api_type column"""
+
+    return db.session.query(UserArtist).filter((UserArtist.api_type==api_type), (UserArtist.user_id==user_id)).all()
+
+# def count_genres_by_user_artists(user_id):
+#     """returns {'Leon Bridges': 2, 'Lady Lamb': 5, 'Tash Sultana': 1, 'Big Wild': 5, 
+#     'Sylvan Esso': 6,...}"""
+
+#     user_join = User.query.options( 
+#              db.joinedload('artists') # attribute for user 
+#                .joinedload('genres')  # attribute from artist
+#             ).get(user_id)  # test is the user id ()   
+
+#     artist_genres = {}
+#     for artist in user_join.artists:
+#         artist_genres[artist.artist_name] = artist_genres.get(artist.artist_name, 0) + len(artist.genres)
+
+#     return artist_genres
+
+# def get_repeating_artists(user_id):
+#     """return list of artists that have multiple genres"""
+
+#     artist_genres = count_genres_by_user_artists(user_id)
+
+#     repeating_artists = []
+#     for artist in artist_genres:
+#         if artist_genres[artist] > 1:
+#             repeating_artists.append(artist)
+
+#     return repeating_artists
+
+
+def optimize_genres(user_id, api_type):
+    """Returns a list of artists by genre with artists in their genre with the 
+    highest count of artists by user
+
+    >>>optimize_genres("test")
+    >>>{'shamanic': ['Beautiful Chorus', 'Rising Appalachia', 'Ayla Nereo'], 
+    'brain waves': ['Alpha Brain Waves'], 'electropop': ['Grimes', 'Sylvan Esso', 
+    'Overcoats', 'Purity Ring', 'Charlotte Day Wilson'],...}"""
+
+    # user_join = User.query.options( 
+    #          db.joinedload('artists') # attribute for user 
+    #            .joinedload('genres')  # attribute from artist
+    #         ).get(user_id)  
+
+    filtered_artists = db.session.query(Artist).filter((UserArtist.api_type==api_type), 
+                    (UserArtist.user_id==user_id), 
+                    (UserArtist.artist_id==Artist.artist_id)).all()
+
+    #count of artists in each genre
+    """{'chillwave': 3, 'dance pop': 3, 'electropop': 6, 
+    'escape room': 4,....}"""
+    artist_genres = count_user_artists_by_genre(user_id) 
```

Comment:
nit: maybe `genre_to_num_artists` or `num_artists_for_genre`. Think about how much more clearly line 242 would read if one of these were the name